# Адаптивный алгоритм интегрирования с помощью формул Ньютона-Котеса

In [3]:
import math

In [56]:
#формула пярмоугольников
def Integrate(h, f, a, b):
    value = 0
    i = 0
    while(abs(a + h * i - b) > 1e-10):
        value += f(a + h * i + h / 2) 
        i += 1
    return value * h

#реализация алгоритма
def solve(a, b, f, epsA = 1e-6, epsO = 1e-6):
    alpha, beta = a, b
    I = E = 0
    while alpha < beta:
        h = beta - alpha
        I_h, I_hdiv2 = Integrate(h, f, alpha, beta), Integrate(h / 2, f, alpha, beta)
        delta = (I_hdiv2 - I_h) / (2**3 - 1) # p = 3
        if abs(delta) < max(epsA, epsO * abs(I)) * h / (b - a):
            alpha = beta
            beta = b
            I = I + I_hdiv2 + delta
            E = E + delta
        else:
            beta = (alpha + beta) / 2
    return I, E

In [57]:
a, b = 0, 2 * math.pi
f = lambda x: math.cos(x) * math.exp(x)
solve(a, b, f)

(267.24581145129923, 1.223414270334899e-05)

Как и ожидалось $$\int_0^{2\pi} e^x \cos x dx = e^{\pi} \sinh \pi \approx 267.25$$

In [58]:
math.exp(math.pi) * math.sinh(math.pi)

267.24582776238236